README.md
# Floating Ribbon

[![pub package](https://img.shields.io/badge/pub-v1.0.0-blue)](https://pub.dev/packages/floating_ribbon) [![Build Status](https://travis-ci.org/101Loop/Floating-Ribbon.svg?branch=master)](https://travis-ci.org/101Loop/Floating-Ribbon) [![LICENSE](https://img.shields.io/badge/license-MIT-green)](https://github.com/101Loop/Floating-Ribbon/blob/master/LICENSE)

A new Flutter package for creating floating ribbons on images.

<p>
    <img src="https://github.com/101Loop/Floating-Ribbon/blob/master/example.gif?raw=true"/>
</p>

## Dependency
```
dependencies:
  floating_ribbon: any
```

## How To Use

> In order to achieve the shimmering effect, you need to add a package in pubspec.yaml.

```
dependencies:
  skeleton_text: any
```

> The code to reproduce it,

```dart
import 'package:skeleton_text/skeleton_text.dart';
import 'package:floating_ribbon/floating_ribbon.dart';
```

```dart
FloatingRibbon(
  height: 85,
  width: 85,
  childHeight: 75,
  childWidth: 75,
  child: Padding(
    padding: const EdgeInsets.all(8.0),
    child: FlutterLogo(),
  ),
  childDecoration: BoxDecoration(color: Colors.grey),
  ribbon: SkeletonAnimation(
    child: Center(
      child: Text(
        'Exclusive',
        style: TextStyle(
          color: Colors.white,
        ),
        textAlign: TextAlign.center,
      ),
    ),
  ),
  shadowHeight: 5,
);

```

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [1.0.0] - 16/08/2019

* Added Example
* Added Documentation

## [0.0.1] - 16/08/2019

* Initial Release


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_test/flutter_test.dart';
import '../lib/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());
  });
}


test/floating_ribbon_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/material.dart';
import 'package:floating_ribbon/floating_ribbon.dart';

/// Tests Passed!
void main() {
  testWidgets('Floating ribbon can be constructed', (tester) async {
    await tester.pumpWidget(
      FloatingRibbon(
        height: 85,
        width: 85,
        childHeight: 75,
        childWidth: 75,
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: FlutterLogo(),
        ),
        ribbon: Text(
          'Test',
          style: TextStyle(
            color: Colors.white,
          ),
          textAlign: TextAlign.center,
        ),
        childDecoration: BoxDecoration(
          color: Colors.grey,
        ),
      ),
    );
  });
}


lib/floating_ribbon.dart
///
/// * author: Mahen Gandhi
/// * email: mahengandhi19@gmail.com
/// * github: @imlegend19
/// * community: 101Loop [https://101loop.com]
///
/// A Flutter package providing an easy way to add floating ribbon to images.
///

library floating_ribbon;

import 'package:flutter/material.dart';
import 'clippers/equilateral_triangle.dart';
import 'clippers/right_triangle.dart';

///
/// An enum defines all supported shadow configurations.
///
/// * [Clipper.right] right triangle shadow
/// * [Clipper.equilateral] equilateral triangle shadow
///

enum Clipper {
  right,
  equilateral,
}

class FloatingRibbon extends StatelessWidget {
  /// Creates a [FloatingRibbon] widget.
  /// [FloatingRibbon] is a widget that displays a horizontal ribbon over [child].
  ///
  /// If the [ribbonTextStyle] argument is null, the text will use the style from the
  /// closest enclosing [DefaultTextStyle].
  ///
  FloatingRibbon({
    Key key,
    @required this.height,
    @required this.width,
    @required this.childHeight,
    @required this.childWidth,
    @required this.child,
    this.childDecoration,
    this.ribbon,
    this.ribbonHeight = 20,
    this.clipper = Clipper.right,
    this.ribbonSwatch = Colors.redAccent,
    this.ribbonShadowSwatch = Colors.red,
    this.shadowHeight,
    this.equilateralTriangleWidth = 5,
  })  : assert(ribbon != null,
            'A non-null child Widget must be provided to a FloatingRibbon widget.'),
        super(key: key);

  /// [height] defines the height of the box occupied by its children.
  ///
  /// Note: [height] >= [childHeight].
  final double height;

  /// [width] defines the width of the box occupied by its children.
  ///
  /// Note: Make sure your [width] > [childWidth] in order to get better results.
  final double width;

  /// [childHeight] defines the height of you [child] which is typically an
  /// Image Widget.
  ///
  /// Note: [childHeight] <= [height]
  final double childHeight;

  /// [childWidth] defines the width of you [child] which is typically an
  /// Image Widget.
  ///
  /// Note: [childWidth] < [width] always. This is because the ribbon width has
  ///       has to be more than the Image Widget width in order to give an
  ///       alluring effect to the ribbon shadow.
  final double childWidth;

  /// A widget renders the ribbon over [child] widget tree.
  /// [child] is the area occupied by the ribbon. You can build [child]
  /// from whatever [Widget] you like.
  ///
  /// * Suggested [child] ->
  ///   Image(
  ///     image: AssetImage('assets/image.png'),
  ///   ),
  final Widget child;

  /// [childDecoration] sets the decoration for its [child]
  ///
  /// Optional [childDecoration] for the child. If the child is an Image you can
  /// wrap it with [ClipRect], [ClipRRect], or [ClipPath] and set the decoration.
  final Decoration childDecoration;

  /// [ribbon] defines the child widget displayed on the ribbon.
  final Widget ribbon;

  /// If non-null, [clipper] defines what type of shadow you want.
  ///
  /// Default value for [clipper] is [Clipper.right].
  final Clipper clipper;

  /// If non-null, [ribbonHeight] defines the height of the image.
  ///
  /// Default value for the [ribbonHeight] is 20. Typically, this should be
  /// 1/4 of [childHeight].
  final double ribbonHeight;

  /// If non-null, [ribbonSwatch] defines the primary swatch for the ribbon.
  ///
  /// Default value for [ribbonSwatch] is Colors.redAccent. Note that the color
  /// should be lightened for the ribbon and darkened for the shadow to add
  /// a charming effect
  final Color ribbonSwatch;

  /// If non-null, [ribbonShadowSwatch] defines the primary swatch for the ribbon.
  ///
  /// Default value for [ribbonShadowSwatch] is Colors.red.
  final Color ribbonShadowSwatch;

  /// If non-null, [shadowHeight] defines the height of the shadow.
  ///
  /// [shadowHeight] defaults to 5 and is an optional parameter. Note that this
  /// is an extra height given to the shadow.
  ///
  /// The formula for shadow height is -> ([width] - [childWidth]) / 2 + [shadowHeight]
  final double shadowHeight;

  /// If non-null, [equilateralTriangleWidth] defines the width of the shadow.
  ///
  /// Note that this would only affect when [clipper] is set to [Clipper.equilateral].
  /// The formula for shadow width is -> ([width] - [childWidth]) / 2 + [equilateralTriangleWidth]
  ///
  /// [equilateralTriangleWidth] defaults to 5.
  final double equilateralTriangleWidth;

  @override
  Widget build(BuildContext context) {
    return Container(
      key: key,
      height: height,
      width: width,
      child: Stack(
        alignment: Alignment.center,
        children: <Widget>[
          Container(
            width: childWidth,
            height: childHeight,
            child: child,
            decoration: childDecoration,
          ),
          Positioned(
            top: childHeight * 0.1,
            child: Column(
              textDirection: TextDirection.ltr,
              children: <Widget>[
                Container(
                  width: width,
                  height: ribbonHeight,
                  color: ribbonSwatch,
                  child: Directionality(
                      textDirection: TextDirection.ltr, child: ribbon),
                ),
                Container(
                  width: width,
                  height: ribbonHeight,
                  child: Row(
                    textDirection: TextDirection.ltr,
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    mainAxisSize: MainAxisSize.max,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: _getShadow(),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> _getShadow() {
    if (clipper == Clipper.right) {
      return <Widget>[
        ClipPath(
          clipper: LeftTriangleClipper(),
          child: Container(
            height: (width - childWidth) / 2 + 5,
            width: (width - childWidth) / 2,
            color: ribbonShadowSwatch,
          ),
        ),
        ClipPath(
          clipper: RightTriangleClipper(),
          child: Container(
            height: (width - childWidth) / 2 + 5,
            width: (width - childWidth) / 2,
            color: ribbonShadowSwatch,
          ),
        ),
      ];
    } else {
      return <Widget>[
        ClipPath(
          clipper: EquilateralTriangleClipper(),
          child: Container(
            height: (width - childWidth) / 2 + shadowHeight,
            width: (width - childWidth) / 2 + equilateralTriangleWidth,
            color: ribbonShadowSwatch,
          ),
        ),
        ClipPath(
          clipper: EquilateralTriangleClipper(),
          child: Container(
            height: (width - childWidth) / 2 + shadowHeight,
            width: (width - childWidth) / 2 + equilateralTriangleWidth,
            color: ribbonShadowSwatch,
          ),
        ),
      ];
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:floating_ribbon/floating_ribbon.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
      debugShowMaterialGrid: false,
      debugShowCheckedModeBanner: false,
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  List<String> _hotels = [
    'Samudra Hotel',
    'Naivedyam',
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey[200],
      appBar: AppBar(
        title: Text('Floating Ribbon Example'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          children: <Widget>[
            Container(
              color: Colors.white,
              height: 40,
              width: MediaQuery.of(context).size.width,
              child: Center(
                child: Text(
                  'clipper = Clipper.right',
                  style: TextStyle(
                    color: Colors.black54,
                    fontSize: 20,
                  ),
                  textAlign: TextAlign.center,
                ),
              ),
            ),
            ListView(
              shrinkWrap: true,
              children: <Widget>[
                restaurantListTile(_hotels.elementAt(0), true),
                restaurantListTile(_hotels.elementAt(1), true),
              ],
            ),
            Container(
              color: Colors.white,
              height: 40,
              width: MediaQuery.of(context).size.width,
              child: Center(
                child: Text(
                  'clipper = Clipper.equilateral',
                  style: TextStyle(
                    color: Colors.black54,
                    fontSize: 20,
                  ),
                  textAlign: TextAlign.center,
                ),
              ),
            ),
            ListView(
              shrinkWrap: true,
              children: <Widget>[
                restaurantListTile(_hotels.elementAt(0), false),
                restaurantListTile(_hotels.elementAt(1), false),
              ],
            ),
          ],
        ),
      ),
    );
  }

  Widget restaurantListTile(String hotel, bool isRight) {
    double width = MediaQuery.of(context).size.width;
    width = width * 0.30;

    return Container(
      decoration: BoxDecoration(
        borderRadius: BorderRadius.all(
          Radius.circular(10.0),
        ),
        color: Colors.white70,
      ),
      margin: EdgeInsets.symmetric(horizontal: 10.0, vertical: 6.0),
      child: Container(
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.start,
          children: <Widget>[
            FloatingRibbon(
              height: 85,
              width: 85,
              childHeight: 75,
              childWidth: 75,
              child: Padding(
                padding: const EdgeInsets.all(8.0),
                child: FlutterLogo(),
              ),
              childDecoration: BoxDecoration(color: Colors.grey),
              ribbon: Center(
                child: Text(
                  'Exclusive',
                  style: TextStyle(
                    color: Colors.white,
                  ),
                  textAlign: TextAlign.center,
                ),
              ),
              clipper: isRight ? Clipper.right : Clipper.equilateral,
              equilateralTriangleWidth: isRight ? null : 5,
              shadowHeight: 7,
            ),
            Column(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.max,
              children: <Widget>[
                Padding(
                  padding: const EdgeInsets.only(left: 15.0, bottom: 5.0),
                  child: Container(
                    width: MediaQuery.of(context).size.width * 0.6,
                    child: Text(
                      hotel,
                      style: TextStyle(
                        color: Colors.black,
                        fontSize: 15,
                        fontWeight: FontWeight.bold,
                        fontFamily: 'Avenir-Bold',
                      ),
                      maxLines: 3,
                    ),
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(left: 15.0),
                  child: Row(
                    children: <Widget>[
                      Padding(
                        padding: const EdgeInsets.only(right: 5.0),
                        child: Icon(Icons.access_alarms,
                            color: Colors.black54, size: 20),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(top: 2.5),
                        child: Text(
                          "30 min",
                          style: TextStyle(
                            color: Colors.black45,
                            fontFamily: 'Avenir-Black',
                            fontSize: 13,
                            fontWeight: FontWeight.w700,
                          ),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                          left: 5.0,
                          right: 5.0,
                          top: 3.0,
                        ),
                        child: Icon(
                          Icons.fiber_manual_record,
                          color: Colors.black54,
                          size: 8.0,
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(top: 3.5),
                        child: Text(
                          'Open',
                          style: TextStyle(
                            color: Colors.green,
                            fontFamily: 'Avenir-Black',
                            fontWeight: FontWeight.w700,
                            fontSize: 13,
                          ),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(
                            left: 5.0, right: 5.0, top: 3.0),
                        child: Icon(Icons.fiber_manual_record,
                            color: Colors.black54, size: 8.0),
                      ),
                      Padding(
                        padding: const EdgeInsets.only(top: 3.0),
                        child: Text(
                          '\$\$',
                          style: TextStyle(
                            color: Colors.black54,
                            fontSize: 13,
                            fontWeight: FontWeight.w900,
                          ),
                        ),
                      )
                    ],
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
